src/draw/stm32_dma2d/lv_gpu_stm32_dma2d.c

```diff
@@ -113,153 +88,560 @@ void lv_draw_stm32_dma2d_ctx_deinit(lv_disp_drv_t * drv, lv_draw_ctx_t * draw_ct
+    assert_param(drawBufferLength % CACHE_ROW_SIZE == 0); // critical
+    assert_param((uint32_t)draw_ctx->buf % CACHE_ROW_SIZE == 0); // critical
+
+    // For performance reasons, both source buffer start address and buffer size *should* be 32-byte aligned since source buffer cache is being cleaned.
+    uint32_t srcBufferLength = lv_area_get_size(src_area) * sizeof(
+                                   lv_color_t); // TODO: verify src pixel size = sizeof(lv_color_t)
+    //assert_param(srcBufferLength % CACHE_ROW_SIZE == 0); // FIXME: assert fails (performance, non-critical)
+    //assert_param((uint32_t)src % CACHE_ROW_SIZE == 0); // FIXME: assert fails (performance, non-critical)
+
+    if(lv_img_src_get_type(src) == LV_IMG_SRC_VARIABLE) {
+        lv_area_t draw_area;
+        if(!_lv_area_intersect(&draw_area, src_area, draw_ctx->clip_area)) return LV_RES_OK;
+        const lv_img_dsc_t * img = src;
+
+        if(img->header.cf == LV_IMG_CF_RGBA8888 && dsc->angle == 0 && dsc->zoom == 256) {
+            // note: LV_IMG_CF_RGBA8888 is actually ARGB8888
+            lv_coord_t dest_stride = lv_area_get_width(draw_ctx->buf_area);
+            lv_point_t src_offset; // source image offset in relation to draw_area
+            src_offset.x = draw_area.x1 - src_area->x1;
+            src_offset.y = draw_area.y1 - src_area->y1;
+            lv_area_move(&draw_area, -draw_ctx->buf_area->x1, -draw_ctx->buf_area->y1);
+            // TODO: It is assumed that img->data buffer is of lv_color_t type. This assumption may be incorrect.
+            _lv_draw_stm32_dma2d_blend_map(draw_ctx->buf, dest_stride, &draw_area, (lv_color_t *)img->data, img->header.w,
+                                           &src_offset, dsc->opa, true);
+            return LV_RES_OK;
+        }
+    }
+
+    return LV_RES_INV;
+}
 
-    lv_draw_sw_img_decoded(draw_ctx, dsc, coords, map_p, color_format);
+void lv_gpu_stm32_dma2d_wait_cb(lv_draw_ctx_t * draw_ctx)
+{
+    lv_disp_t * disp = _lv_refr_get_disp_refreshing();
+    _lv_gpu_stm32_dma2d_await_dma_transfer_finish(disp->driver);
+    lv_draw_sw_wait_for_finish(draw_ctx);
 }
 
-static void lv_draw_stm32_dma2d_blend_fill(lv_color_t * dest_buf, lv_coord_t dest_stride, const lv_area_t * fill_area,
-                                           lv_color_t color)
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
+/**
+ * @brief Fills draw_area with specified color.
+ * @param color color to be painted, note: alpha is ignored
+ */
+LV_STM32_DMA2D_STATIC void _lv_draw_stm32_dma2d_blend_fill(const lv_color_t * dest_buf, lv_coord_t dest_stride,
+                                                           const lv_area_t * draw_area, lv_color_t color, lv_opa_t opa)
 {
-    /*Simply fill an area*/
-    int32_t area_w = lv_area_get_width(fill_area);
-    int32_t area_h = lv_area_get_height(fill_area);
-    invalidate_cache();
+    assert_param(!isDma2dInProgess); // critical
```

Comment:
Why assert here?

Reply:
This assert is to ensure that any previous dma2d transfer initiated within this driver (only) was properly awaited for and finished.